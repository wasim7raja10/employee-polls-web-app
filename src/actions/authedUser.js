const LOGIN_AUTHED_USER = 'LOGIN_AUTHED_USER'
const LOGOUT_AUTHED_USER = 'LOGOUT_AUTHED_USER'

const loginAuthedUser = (authedUser) => {
  return {
    type: LOGIN_AUTHED_USER,
    authedUser,
  }
}

const logoutAuthedUser = () => {
  return {
    type: LOGOUT_AUTHED_USER,
  }
}

const handleLoginAuthedUser = (credential) => {
  return (dispatch, getState) => {
    const { users } = getState()
    const user = Object.values(users).find(user => user.id === credential.username && user.password === credential.password)
    if (!!user) {
      return dispatch(loginAuthedUser(user))
    } else {
      return false
    }
  }
}

const handleLogoutAuthedUser = () => {
  return (dispatch) => {
    dispatch(logoutAuthedUser())
  }
}

export { LOGIN_AUTHED_USER, LOGOUT_AUTHED_USER, loginAuthedUser, logoutAuthedUser, handleLoginAuthedUser, handleLogoutAuthedUser }