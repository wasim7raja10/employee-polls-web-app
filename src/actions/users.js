export const RECEIVE_USERS = "RECEIVE_USERS";
export const ADD_USER_QUESTION = "ADD_USER_QUESTION";
export const ADD_USER_ANSWER = "ADD_USER_ANSWER";

const addUserQuestion = ({ author, id }) => {
  return {
    type: ADD_USER_QUESTION,
    author,
    id
  };
}

const addUserAnswer = ( authUser, qid, answer ) => {
  return {
    type: ADD_USER_ANSWER,
    authUser,
    qid,
    answer,
  };
}

const receiveUsers = (users) => {
  return {
    type: RECEIVE_USERS,
    users,
  };
}

export { addUserQuestion, addUserAnswer, receiveUsers }