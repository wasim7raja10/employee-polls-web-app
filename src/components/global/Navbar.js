import { connect } from "react-redux"
import { Link } from "react-router-dom"
import { handleLogoutAuthedUser } from "../../actions/authedUser"

const Navbar = ({ dispatch, authUser }) => {
  const logoutHandle = () => {
    dispatch(handleLogoutAuthedUser())
  }
  return (
    <div className="navbar bg-base-100">
      <div className="flex-1">
        <Link data-testid="logo" to={"/"} className="btn btn-ghost normal-case text-xl">Employee Poll</Link>
        <h2 data-testid="authUserName" className=" w-full text-center text-xl">{authUser?.name}</h2>
      </div>
      <div className="flex-none">
        <ul data-testid="navbuttons" className="menu menu-horizontal px-1">
          <li><Link data-testid="home-link" to={"/"}>Home</Link></li>
          <li><Link data-testid="new-poll-link" to={"/add"}>New Poll</Link></li>
          <li><Link data-testid="leaderboard-link" to={"/leaderboard"}>Leaderboard</Link></li>
          {authUser ? <li><button data-testid="logout-link" onClick={logoutHandle}>Logout</button></li>
            : <li><Link data-testid="login-link" to={"/login"}>Login</Link></li>}
        </ul>
      </div>
    </div>
  )
}

const mapStateToProps = ({ authUser }) => ({
  authUser,
})

export default connect(mapStateToProps)(Navbar)