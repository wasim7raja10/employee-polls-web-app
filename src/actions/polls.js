import { saveQuestion, saveQuestionAnswer } from "../data/api";
import { addUserAnswer, addUserQuestion } from "./users";

export const ADD_QUESTION = "ADD_QUESTION";
export const ADD_ANSWER_QUESTION = "ADD_ANSWER_QUESTION";
export const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";

const addQuestion = (question) => {
  return {
    type: ADD_QUESTION,
    question,
  };
}

const addAnswerQuestion = (authUser, qid, answer) => {
  return {
    type: ADD_ANSWER_QUESTION,
    authUser,
    qid,
    answer,
  };
}

const receiveQuestions = (questions) => {
  return {
    type: RECEIVE_QUESTIONS,
    questions,
  };
}

const handleAddQuestion = (firstOption, secondOption) => {
  return async (dispatch, getState) => {
    const { authUser } = getState();
    const question = {
      optionOneText: firstOption,
      optionTwoText: secondOption,
      author: authUser,
    };
    const questionResponse = await saveQuestion(question);
    dispatch(addQuestion(questionResponse));
    dispatch(addUserQuestion(questionResponse));
  };
}

const handleAddAnswer = (questionId, answer) => {
  return async (dispatch, getState) => {
    const { authUser } = getState();
    const answerObject = {
      authedUser: authUser.id,
      qid: questionId,
      answer,
    };
    const isAdded = await saveQuestionAnswer(answerObject);
    if(isAdded) {
      dispatch(addAnswerQuestion(authUser.id, questionId, answer));
      dispatch(addUserAnswer(authUser.id, questionId, answer));
    }
  };
}

export { handleAddQuestion, handleAddAnswer, receiveQuestions };