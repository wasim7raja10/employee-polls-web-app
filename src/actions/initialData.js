import { getInitialData } from "../data/api";
import { receiveUsers } from "./users";
import { receiveQuestions } from "./polls";

const receiveInitialData = () => {
  return async (dispatch) => {
    const { users, questions } = await getInitialData();
    dispatch(receiveUsers(users));
    dispatch(receiveQuestions(questions));
  };
}

export { receiveInitialData };
