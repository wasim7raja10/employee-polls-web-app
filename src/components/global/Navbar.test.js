import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import React from "react";
import Navbar from "./Navbar";
import { configureStore } from "@reduxjs/toolkit";
import reducer from "../../reducers";
import { loginAuthedUser } from "../../actions/authedUser";

const store = configureStore({ reducer })

describe("Navbar", () => {
  it("should render the component", () => {
    store.dispatch(loginAuthedUser({
      username: "tylermcginnis", password: ""
    }))
    const view = render(
      <Provider store={store}>
        <BrowserRouter>
          <Navbar />
        </BrowserRouter>
      </Provider>
    );
    expect(view).toBeDefined();
    expect(view).toMatchSnapshot();
  });

  it("should display all elements", () => {
    store.dispatch(loginAuthedUser({
      username: "tylermcginnis", password: ""
    }))
    const view = render(
      <Provider store={store}>
        <BrowserRouter>
          <Navbar />
        </BrowserRouter>
      </Provider>
    );

    const homeLinkElement = view.getByTestId("home-link");
    const newPollLinkElement = view.getByTestId("new-poll-link");
    const leaderboardLinkElement = view.getByTestId("leaderboard-link");
    const logoutLinkElement = view.getByTestId("logout-link");

    expect(homeLinkElement.textContent).toBe("Home");
    expect(newPollLinkElement.textContent).toBe("New Poll");
    expect(leaderboardLinkElement.textContent).toBe("Leaderboard");
    expect(logoutLinkElement.textContent).toBe("Logout");
  });
});
