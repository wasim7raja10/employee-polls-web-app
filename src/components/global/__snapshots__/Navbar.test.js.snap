// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar should render the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="navbar bg-base-100"
      >
        <div
          class="flex-1"
        >
          <a
            class="btn btn-ghost normal-case text-xl"
            data-testid="logo"
            href="/"
          >
            Employee Poll
          </a>
          <h2
            class=" w-full text-center text-xl"
            data-testid="authUserName"
          />
        </div>
        <div
          class="flex-none"
        >
          <ul
            class="menu menu-horizontal px-1"
            data-testid="navbuttons"
          >
            <li>
              <a
                data-testid="home-link"
                href="/"
              >
                Home
              </a>
            </li>
            <li>
              <a
                data-testid="new-poll-link"
                href="/add"
              >
                New Poll
              </a>
            </li>
            <li>
              <a
                data-testid="leaderboard-link"
                href="/leaderboard"
              >
                Leaderboard
              </a>
            </li>
            <li>
              <button
                data-testid="logout-link"
              >
                Logout
              </button>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="navbar bg-base-100"
    >
      <div
        class="flex-1"
      >
        <a
          class="btn btn-ghost normal-case text-xl"
          data-testid="logo"
          href="/"
        >
          Employee Poll
        </a>
        <h2
          class=" w-full text-center text-xl"
          data-testid="authUserName"
        />
      </div>
      <div
        class="flex-none"
      >
        <ul
          class="menu menu-horizontal px-1"
          data-testid="navbuttons"
        >
          <li>
            <a
              data-testid="home-link"
              href="/"
            >
              Home
            </a>
          </li>
          <li>
            <a
              data-testid="new-poll-link"
              href="/add"
            >
              New Poll
            </a>
          </li>
          <li>
            <a
              data-testid="leaderboard-link"
              href="/leaderboard"
            >
              Leaderboard
            </a>
          </li>
          <li>
            <button
              data-testid="logout-link"
            >
              Logout
            </button>
          </li>
        </ul>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
